From b9d82dcf9fceae6552444a56659a34d7d48439f5 Mon Sep 17 00:00:00 2001
From: Aradhya Bhatia <a-bhatia1@ti.com>
Date: Thu, 6 Jul 2023 18:31:34 +0530
Subject: [PATCH 127/143] drm/tidss: Set OLDI clock to bypass-25MHz during
 probe

This is being done because the AM625 bootloaders set up the OLDI PLL
frequency to (pixel-clock x 7) value for splash screen functionality.
The enable_oldi bit is also set during this time, which makes the DSS
receive the exact pixel-clock, (pixel-clock * 7)/7.

During linux tidss probe, the DSS controller gets soft reset and clears
the enable_oldi bit, but not the OLDI PLL. This makes the controller
receive (pixel-clock * 7) frequency since the clock divider is only
activated when enable_oldi bit is set.

For usecases where OLDI DT overlay is not applied, the enable_oldi bit
is never going to be set. In these cases, receiving such a high clock
can potentially cause damage to the DSS controller hardware.

Set AM625 OLDI clock to the bypass PLL frequency, i.e. 25MHz to fix the
issue.

Signed-off-by: Aradhya Bhatia <a-bhatia1@ti.com>
Reviewed-by: Devarsh Thakkar <devarsht@ti.com>
Tested-by: Jayesh Choudhary <j-choudhary@ti.com>
Signed-off-by: Vignesh Raghavendra <vigneshr@ti.com>
---
 drivers/gpu/drm/tidss/tidss_dispc.c | 12 ++++++++++++
 drivers/gpu/drm/tidss/tidss_drv.h   |  2 ++
 2 files changed, 14 insertions(+)

diff --git a/drivers/gpu/drm/tidss/tidss_dispc.c b/drivers/gpu/drm/tidss/tidss_dispc.c
index c95fbeea8291..307db1e47ee1 100644
--- a/drivers/gpu/drm/tidss/tidss_dispc.c
+++ b/drivers/gpu/drm/tidss/tidss_dispc.c
@@ -2988,6 +2988,18 @@ int dispc_init(struct tidss_device *tidss)
 		}
 		dispc->vp_clk[i] = clk;
 
+		if (dispc_get_output_type(dispc, i) == DISPC_OUTPUT_OLDI &&
+		    feat->subrev == DISPC_AM625) {
+			r = clk_set_rate(dispc->vp_clk[i],
+					 TIDSS_AM625_IDLE_OLDI_CLOCK);
+			if (r) {
+				dev_err(dev,
+					"vp%d: failed to set oldi clk rate to %u\n",
+					i, TIDSS_AM625_IDLE_OLDI_CLOCK);
+				return r;
+			}
+		}
+
 		gamma_table = devm_kmalloc_array(dev, gamma_size,
 						 sizeof(*gamma_table),
 						 GFP_KERNEL);
diff --git a/drivers/gpu/drm/tidss/tidss_drv.h b/drivers/gpu/drm/tidss/tidss_drv.h
index 7919c5a51890..2bf4088051ae 100644
--- a/drivers/gpu/drm/tidss/tidss_drv.h
+++ b/drivers/gpu/drm/tidss/tidss_drv.h
@@ -21,6 +21,8 @@
 /* For DSSes with 2 OLDI TXes */
 #define TIDSS_MAX_BRIDGES_PER_PIPE	2
 
+#define TIDSS_AM625_IDLE_OLDI_CLOCK 25000000
+
 typedef u32 dispc_irq_t;
 
 struct tidss_device {
-- 
2.41.0

